Improve explanation for adding a new block device (#61)
ARMmbed/mbed-os-example-blockdevice: README.md

for suggestions on what could be wrong and how to fix it.

## Changing the block device

In Mbed OS, a C++ classes that inherits from the [BlockDevice](https://os.mbed.com/docs/latest/reference/storage.html#block-devices)
interface represents each block device. You can change the filesystem in the
example by changing the class declared in main.cpp.

``` diff
-SPIFBlockDevice bd(
-        MBED_CONF_SPIF_DRIVER_SPI_MOSI,
-        MBED_CONF_SPIF_DRIVER_SPI_MISO,
-        MBED_CONF_SPIF_DRIVER_SPI_CLK,
-        MBED_CONF_SPIF_DRIVER_SPI_CS);

> +SDBlockDevice bd(

Reviewer: I liked the diff in the original file. Could you put it back?

> Each device is represented by a C++ class that inherits from the interface class [BlockDevice](https://os.mbed.com/docs/latest/reference/storage.html#block-devices). These classes take their default configuration from the component configuration file. This may be found in `/mbed-os/storage/blockdevice/` under the path corresponding to the block device type—for example [mbed_lib.json](https://github.com/ARMmbed/mbed-os/blob/master/storage/blockdevice/COMPONENT_SPIF/mbed_lib.json). 

Reviewer: The first link should be to the BlockDevice API

> In this example, you can determine which block device is used by modifying the declaration of the object `bd` in main.cpp. For instance, if you wanted to add a SPIF block device you would need to declare `bd` as an object of the type SPIFBlockDevice.

Reviewer: ```suggestion
In this example, you can determine which block device is used by modifying the type of `bd` in main.cpp. For instance, if you wanted to use a SPIF block device you would declare `bd` as an SPIFBlockDevice instance.
```

```

> Then, you would override the default device settings in `mbed_app.json`, defining any pin reassignments here:

Reviewer: You previsouly explained the two step process: 
1- Check whether you need to add the component
2- Possibly modify the pin assignment
I think that it was a good explanation

for suggestions on what could be wrong and how to fix it.

## Changing the block device

> Mbed-OS supports a variety of block device types, more information on supported devices can be found [here](https://os.mbed.com/docs/mbed-os/v6.5/apis/data-storage.html#Default-BlockDevice-configuration).

Reviewer: The link should point to "latest" rather than a specific Mbed OS version
```suggestion
Mbed-OS supports a variety of block device types, more information on supported devices can be found [here](https://os.mbed.com/docs/mbed-os/latest/apis/data-storage.html#Default-BlockDevice-configuration).
```

In Mbed OS, a C++ classes that inherits from the [BlockDevice](https://os.mbed.com/docs/latest/reference/storage.html#block-devices)

> interface represents each block device. You can change the filesystem in the

Reviewer: This information on changing the block device by changing the class declared in main.cpp should have been kept.

> Each device is represented by a C++ class that inherits from the super class [BlockDevice](https://os.mbed.com/docs/latest/reference/storage.html#block-devices). These classes take their default configuration from the component configuration file. This may be found in `/mbed-os/storage/blockdevice/` under the path corresponding to the block device type. The default settings can be overridden in either this file, or, in `mbed_app.json`.

Reviewer: What is a super class? Call this an interface class instead.

Author: A super class is the parent class of an object. I hadn't noticed that there is no actual implementation of BlockDevice!

> Each device is represented by a C++ class that inherits from the super class [BlockDevice](https://os.mbed.com/docs/latest/reference/storage.html#block-devices). These classes take their default configuration from the component configuration file. This may be found in `/mbed-os/storage/blockdevice/` under the path corresponding to the block device type. The default settings can be overridden in either this file, or, in `mbed_app.json`.

Reviewer: I'd suggest to give an example of the component file.
```suggestion
Each device is represented by a C++ class that inherits from the super class [BlockDevice](https://os.mbed.com/docs/latest/reference/storage.html#block-devices). These classes take their default configuration from the component configuration file. This may be found in `/mbed-os/storage/blockdevice/` under the path corresponding to the block device type, for example `https://github.com/ARMmbed/mbed-os/blob/master/storage/blockdevice/COMPONENT_SPIF/mbed_lib.json`.
```

> Each device is represented by a C++ class that inherits from the super class [BlockDevice](https://os.mbed.com/docs/latest/reference/storage.html#block-devices). These classes take their default configuration from the component configuration file. This may be found in `/mbed-os/storage/blockdevice/` under the path corresponding to the block device type. The default settings can be overridden in either this file, or, in `mbed_app.json`.

Reviewer: "The default settings can be overridden in either this file, or, in `mbed_app.json`"
We should not recommend changing the component configuration file.

> For instance, to add a SPI flash block device to an STM32F29ZI board, the following modifications are made to the application configuration file.

Reviewer: ```suggestion
For instance, to add a SPIF block device to an NUCLEO_F429ZI board, modify `mbed_app.json` as follows:
```

```
   "target_overrides": {
         ...
         "NUCLEO_F429ZI": {
             "target.components_add": ["SPIF"],
         },
         ...
     }
```

> Then, if the pin assignments do not match, reassignments are added to the component configuration.  

Reviewer: If this applies to the component library file then we should not recommend changing it.

+        MBED_CONF_SD_SPI_CLK,
+        MBED_CONF_SD_SPI_CS);
```

> You may need to make modifications to the application configuration file if you're using a physical storage device that isn't included in your targets default configuration (check in `targets.json` for this). To do this, add your physical storage device as a component in `mbed_app.json` as follows:

Reviewer: ```suggestion
You may need to make modifications to the application configuration file if you're using a physical storage device that isn't included in your target's default configuration (check in `targets.json` for this). To do this, add your physical storage device as a component in `mbed_app.json` as follows:
```

> You may also need to make modifications to the application configuration file if you're using a storage device that isn't part of your targets configuration (check in `targets.json` for this). To do this, add your storage device as a component in `mbed_app.json`, this will override the default block device settings.

Reviewer: ```suggestion
You may also need to make modifications to the application configuration file if you're using a physical storage device that isn't included in your target's default configuration (check `targets.json` for this). To do this, add your physical storage device as a component in `mbed_app.json` as follows:
```

``` json
   "target_overrides": {
         "K64F": {
             "target.components_add": ["SPIF"],
         }
     }
```

> You should also add the pin assignments into the configuration, if they aren't already defined, as follows:

Reviewer: The pin assignment is provided by default in the component's mbed_lib.json. 
```suggestion
You can also modify the pin assignments for your component as follows:
```

> Then, if the pin assignments to the peripheral are inconsistent, define any reassignments here as well.

Reviewer: ```suggestion
You can also modify the pin assignments for your driver as follows:
```

``` json
   "target_overrides": {
         "K64F": {
             "target.components_add": ["SPIF"],
             "spif-driver.SPI_MOSI": "PC_12",
             "spif-driver.SPI_MISO": "PC_11",
             "spif-driver.SPI_CLK":  "PC_10",
             "spif-driver.SPI_CS":   "PA_15"
         }
     }
```

> Alternatively, you may also choose to reimplement the function `BlockDevice::get_default_instance()`, but this is more involved and isn't recommended unless you have a good understanding of block devices and the overall BlockDevice API.

Reviewer: ```suggestion
Alternatively, you may use the system's default block device `BlockDevice *bd = BlockDevice::get_default_instance()` but this will require more code changes to the example.
```

Author: The original text more accurately represents the last option IMO. Here is the explanation taken from BlockDevice.h. 
``` cpp
     * An application can override all target settings by implementing
     * BlockDevice::get_default_instance() - the default
     * definition is weak, and calls get_target_default_instance().
 ```

Reviewer: The example does not currently use BlockDevice::get_default_instance. So instead of instantiating a block device, it could get the default block device. It could also reimplement that function as you suggest, which is yet another step.

Author: I see, it probably is better to mention this at the very start before discussing ways to add non-default devices.

Reviewer: For the sake of simplicity and without going into too much detail, I propose the above. I think this gives a good starting point for someone who would like to change the default implementation which they will know is possible since this is a weak function.

         ...
     }
```
This can also be done through the application configuration file.
 
```
    "target_overrides": {
          ...
          "NUCLEO_F429ZI": {
              "spif-driver.SPI_MOSI": "PC_12",
              "spif-driver.SPI_MISO": "PC_11",
              "spif-driver.SPI_CLK":  "PC_10",
              "spif-driver.SPI_CS":   "PA_15"
          },
          ...
      }
 ```

> Alternatively, `BlockDevice::get_default_instance()` may be re-implemented by the user to bypass all default settings.

Reviewer: We should say that this is a more complicated option. 
The example should also be modified to call this function instead of instantiating a block device directy.
